Search bar

NicoMSK opened the pull request:
- Add a task search to the todo

NicoMSK committed:
feat: Add a search bar

- Add a task search to the todo

NicoMSK committed:
feat: Add to the search input, the ability to clean up with a mouse click

kagegrifon commented on src/scripts/edit/editController.js:
```diff
@@ -54,5 +70,13 @@ export function toggleTaskCompletion(event) {
+  if (renderSearch) {
+    return;
+  };
+  /// нужен совет как лучше сделать, оба варианта мне нравятся и оба работают
```
обсудим

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -95,11 +95,16 @@ function getFilteredTasks() {
-export { todos, addTodo, deleteTodo, getTaskById, getFilteredTasks, setCurrentFilterValue, getCurrentFilterValue, FILTER, FILTER_LABELS, updateTaskProperty };
+function searchTasks(inputValue) {
+  return todos.filter((task) =>
+    task.title.toLocaleLowerCase().includes(inputValue)
```
метод аналогичен фильтру, их бы можно было совместить
я расскажу как

kagegrifon reviewed (changes requested):
хорошо бы сделать, чтобы фильтрация по isCompleted и по тексту работали вместе
надо бы поправить логику, чтобы фильтрация была скрыта в модели

NicoMSK commented on src/scripts/edit/editController.js:
```diff
@@ -2,6 +2,17 @@ import * as editView from './editView.js';
+
+  if (inputValue !== "") {
+    formView.renderList(searchTodos);
+    return true;
```
Поправили

NicoMSK committed:
refactor: Partial correction based on comments

NicoMSK committed:
refactor: Redo the todos filter

 - One single filter has been made in the model, the controller has less code, everything is now checked and filtered in the model

- Filtered and search by task and task status, all through one filter

NicoMSK commented on src/scripts/edit/editController.js:
```diff
@@ -2,6 +2,17 @@ import * as editView from './editView.js';
+
+function renderSearchResults() {
+  const inputValue = searchView.searchInput.value.toLowerCase();
+  const searchTodos = model.searchTasks(inputValue);
```
Переписал всю логику фильтра, теперь он один и для поиска и для статуса

NicoMSK commented on src/scripts/search/searchController.js:
```diff
@@ -0,0 +1,21 @@
+  if (inputValue === "") {
+    formView.renderList(model.getFilteredTasks());
+  } else {
+    model.setCurrentFilterValue(model.FILTER.all);
```
Поправил

kagegrifon commented on src/scripts/dialog/dialogController.js:
```diff
@@ -1,23 +1,42 @@
+function renderFilteredTasksOnDialogClose() {
+  const inputValue = searchView.searchInput.value;
+
+  if (inputValue === "") {
```
а что если сначала ввел что-то, а потом убрал - надо же перерендерить будет

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -94,23 +104,53 @@ function getCurrentFilterValue() {
 };
 
-function getFilteredTasks() {
+function getFilteredTasks(task) {
```
название функции теперь надо поменять
что-нибудь вроде filterTaskByIsComplete

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -94,23 +104,53 @@ function getCurrentFilterValue() {
   };
 };
 
+let normalizedSearchText = null;
```
для исходного значения для normalizedSearchText больше подходит пустая строка, чем null
null используется для сущностей объект и массив

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -94,23 +104,53 @@ function getCurrentFilterValue() {
+  return normalizedSearchText;
+};
+
+function searchTextTask(item) {
```
1) в getFilteredTasks ты передаешь task, а здесь называешь item. Что может ввести в заблуждение, что это разные вещи, хотя это про тоже самое. 

2) Item плохое название, потому что максимально абстрактное. 

3) имя функции не очень подходит. Лучше назвать filterTaskByTitle по аналогии с filterTaskByIsComplete

NicoMSK replied:
1) Да, точно, я хотел потом поправить и забыл
2) ок, понял
3) Переменовал

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -94,23 +104,53 @@ function getCurrentFilterValue() {
+
+export function getTasks() {
+  return todos.filter(task => {
+    const matchesStatus = getFilteredTasks(task);
```
matchesStatus -> isComleteFilterResult

NicoMSK replied:
Ок, переименовал

kagegrifon commented on src/scripts/todos/model.js:
```diff
@@ -94,23 +104,53 @@ function getCurrentFilterValue() {
+export function getTasks() {
+  return todos.filter(task => {
+    const matchesStatus = getFilteredTasks(task);
+    const matchesText = isStringNotEmpty() ? searchTextTask(task) : true;
```
matchesText -> titleFilterResult

тернарник заставляет меня больше думать и переворачивать в голове

можно написать еще так
```
let titleFilterResult = true
if (!isStringNotEmpty()) {
   titleFilterResult = filterTaskByTitle(task)
}
```

NicoMSK replied:
Ок, переписал на if

kagegrifon reviewed (approved):
Переделал как договаривались. Отличная работа!
Есть несколько небольших правок и можно мержить

NicoMSK committed:
refactor: Made changes according to the comments

NicoMSK committed:
refactor: Made fixes according to the comments of the code review
